### ISSUE TITLE
TLSv1.3: Incoming CCS message (SSL3_RT_CHANGE_CIPHER_SPEC) not detected using SSL_set_msg_callback

### ISSUE THREAD

**Issue by martinschmatz** [openssl/openssl#25166]
OpenSSL v3.3.1 (and earlier)

***Issue***

TLSv1.3: 

A TLS v1.3 server can send a CCS (change cipher spec) message to disguise the TLSv1.3 session as a TLS 1.2 session for middle-box compatibility. While not part of the TLSv1.3 standard [[RFC8446](https://datatracker.ietf.org/doc/html/rfc8446)], these CCS messages are in ubiquitous use, e.g. in `s_server`.

When using the following message callback `myMessageCallback` via [SSL_set_msg_callback](https://docs.openssl.org/master/man3/SSL_CTX_set_msg_callback/), outgoing (client -> server) CCS messages are correctly detected, BUT incoming (server -> client) CCS messages are NOT detected: 
```
void myMessageCallback(int write_p, int version,
                       int content_type, const void *buf,
                       size_t len, SSL *ssl, void *arg) {
    // We check for a TLSv1.3 Change Cipher Spec Message
    if (version == TLS1_3_VERSION && content_type == SSL3_RT_CHANGE_CIPHER_SPEC) {
        // Is traffic coming in? 
        if (write_p == 0) {
            //Do something like:
            fprintf(stdout, "Incoming CCS message detected\n");
        }
        // Is traffic being sent out? 
        if (write_p == 1) {
            //Do something like:
            fprintf(stdout, "Outgoing CCS message detected\n");
        }
    }                        
}
```
Remark: The callback correctly sees the `SSL3_RT_HEADER` message preceding the incoming CCS message.

***Root cause***

The reason is that incoming CCS messages are ignored (= qualified as empty records) for TLSv1.3 [here](https://github.com/openssl/openssl/blob/0793071efaa7f61828b555128587db48c5d24962/ssl/record/methods/tls_common.c#L752).

***Remediation***

If one needs to capture such incoming (server -> client) CCS messages for TLSv1.3 in the message callback (e.g. for timing measurements), one can still 'ignore' processing of incoming CCS messages by treating those as empty records, but can nevertheless trigger the callback function (after [the comment](https://github.com/openssl/openssl/blob/0793071efaa7f61828b555128587db48c5d24962/ssl/record/methods/tls_common.c#L752)) like this: 
```
        /*
         * CCS messages are ignored in TLSv1.3. We treat it like an empty
         * handshake record - we nevertheless trigger the callback function if needed
         */
        if (rl->msg_callback)
            rl->msg_callback(0, TLS1_3_VERSION, SSL3_RT_CHANGE_CIPHER_SPEC, 
                             thisrr->data, 1, rl->cbarg);
```

Adding the trigger of the callback function as shown above also changes the output for the built-in `SSL_trace` message callback [here](https://github.com/openssl/openssl/blob/0793071efaa7f61828b555128587db48c5d24962/ssl/t1_trce.c#L1770) from an incomplete feedback:
```
Received TLS Record
Header:
  Version = TLS 1.2 (0x303)
  Content Type = ChangeCipherSpec (20)
  Length = 1
```
to the complete feedback (note the last line):
```
Received TLS Record
Header:
  Version = TLS 1.2 (0x303)
  Content Type = ChangeCipherSpec (20)
  Length = 1
    change_cipher_spec (1)
```

**Comment by mattcaswell** [comment]
Looks correct to me.

> If one needs to capture such incoming (server -> client) CCS messages for TLSv1.3 in the message callback (e.g. for timing measurements), one can still 'ignore' processing of incoming CCS messages by treating those as empty records, but can nevertheless trigger the callback function (after [the comment](https://github.com/openssl/openssl/blob/0793071efaa7f61828b555128587db48c5d24962/ssl/record/methods/tls_common.c#L752)) like this:

Seems like a reasonable fix. Would you like to raise a PR for this?

**Comment by martinschmatz** [comment]
I consider this a minor issue with a 2-line fix --> If it can just be added by the maintainers in one of the upcoming minor releases, that would be great (as it would free me from doing a PR).

**Comment by martinschmatz** [comment]
Btw: Here a super minor thing I stumbled over while searching for the right place for the remediation:

The variable `is_tls13` is set [here](https://github.com/openssl/openssl/blob/2f33265039cdbd0e4589c80970e02e208f3f94d2/ssl/record/rec_layer_s3.c#L623C5-L623C14), not used [here](https://github.com/openssl/openssl/blob/2f33265039cdbd0e4589c80970e02e208f3f94d2/ssl/record/rec_layer_s3.c#L707) - but rather produced again via a call to `SSL_CONNECTION_IS_TLS13`.

**Comment by mattcaswell** [comment]
Looks like similar is true here:

https://github.com/openssl/openssl/blob/2f33265039cdbd0e4589c80970e02e208f3f94d2/ssl/record/rec_layer_s3.c#L947

**Comment by mattcaswell** [comment]
> I consider this a minor issue with a 2-line fix --> If it can just be added by the maintainers in one of the upcoming minor releases, that would be great (as it would free me from doing a PR).

Ok. Fix in #25169

**Comment by martinschmatz** [comment]
Thanks! Appreciated!
